### DISCUSSION TITLE
D1: Relation IDs are lost when modifying table schema

### DISCUSSION THREAD

**Discussion by hrueger** [prisma/prisma#24539] [Q&A] [answered]
### Question

Hi,
I'm not sure if this is a bug or if prisma can work around this.
When I have an optional relation defined (using an D1 DB) with the default `onDelete: SetNull` and generate a migration to update a table, the relation is lost due to the migration deleting the data.

### How to reproduce (optional)

Take the following migration for example:
```
CREATE TABLE "new_Product" (
    "id" TEXT NOT NULL PRIMARY KEY,
    "createdAt" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "updatedAt" DATETIME NOT NULL,
    "name" TEXT NOT NULL,
    "description" TEXT,
    ...some more rows
);
INSERT INTO "new_Product" ("createdAt", "description", "id", "name", "updatedAt", ...) SELECT "createdAt", "description", "id", "name", "updatedAt", ... FROM "Product";
DROP TABLE "Product";
ALTER TABLE "new_Product" RENAME TO "Product";
```
In another table with an optional relation to `Product`, the `productId` column now is `NULL` for every row...
The same problem exists with `ManyToMany` relations...

Questions:
- Is that known and expected?
- Are there workarounds for the Migration? I'd rather avoid updating my schema to use onDelete: Restrict everywhere because then I deleting those entries (explicitely in my code) becomes more complex
 
Thanks in advance!

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by janpio** [comment, marked as answer]
> * Is that known and expected?

No, I don't think we have a bug report issue for this.

Can you open one, optimally, with a reproduction repository that we can use to try this out?

> * Are there workarounds for the Migration? I'd rather avoid updating my schema to use onDelete: Restrict everywhere because then I deleting those entries (explicitely in my code) becomes more complex

From a quick look the problem seems to be the combination of the referential action and the generated migration SQL. Maybe the migration could disable the referential actions? Or we could temporarily change it for the migration? Something along those lines probably.

**Reply by hrueger** [reply]
Wow, thanks for the quick reply. I just did some more research, but unfortunately I found this:

> It is not possible to disable recursive foreign key actions

[Source (bottom of the page, 6.3)](https://www.sqlite.org/foreignkeys.html)

I'll create a bug report, though 👍

**Reply by hrueger** [reply]
Actually, after some investigation and failing to create a reproduction with SQLite, I found that it only happens with D1 :-(
I assumed it was the same for SQLite and wrote about that for the sake of simplicity. Sorry for the confusion ;-)
Working on the updated issue + reproduction with D1.
However, I think Cloudflare should rather fix this than prisma working around it ;-)

**Reply by janpio** [reply]
Not surprised, but definitely something for us to figure out. Maybe the Cloudflare behavior is actually wrong and unexpected, or our normal measure to avoid this usually just does not work there. With a reproduction we'll be able to figure that out.

**Reply by hrueger** [reply]
Just for completeness (and for anyone googling):
for continuing the conversation and a minimal reproduction, please see this issue: https://github.com/prisma/prisma/issues/24540
